### PR TITLE
Intergrate free libva and intel-vaapi-driver into deb package

### PR DESCRIPTION
**Pull request by nyanmisaka**
**Changes**
- Intergrate free [libva](https://github.com/intel/libva) and [intel-vaapi-driver](https://github.com/intel/intel-vaapi-driver) with MIT license into deb package.
(I am not sure whether the format of the changes in debian/copyright is accurate.)

- Avoid downloading unnecessary amd64 headers when building armhf/arm64 arch.

- Drop the unused VDPAU libs which has almost been replaced by NVENC/NVDEC.

- You can uncomment the parts of `gmmlib`, `media-driver`, and `MediaSDK` in `docker-build.sh` to get a deb package that supports all QSV HWA functions.
> It is commented out by default, as the full-featured version is non-free and cannot be distributed. In addition, the transcoding function of the non-full-featured version is incomplete, so it is better to use VAAPI.

**Issues**

https://github.com/jellyfin/jellyfin/issues/2726

https://github.com/jellyfin/jellyfin/issues/2724

https://github.com/jellyfin/jellyfin-ffmpeg/issues/16

### CODE REVIEW

File: debian/control [hunk 1]
```diff
@@ -45,17 +45,29 @@ Build-Depends:
...
+# which will be shipped with jellyfin-ffmpeg package.
+# Vaapi currently only supports Intel and AMD gfx
+# and the drivers they provide only support x86/x64.
+# libva-dev [!armhf !arm64],
+#
 # --enable-vdpau
- libvdpau-dev,
+# Vdpau has almost been replaced by nvenc/nvdec, and it is not required in jellyfin.
+# libvdpau-dev,
+#
```
Comment (JustAMan) [inline debian/control]: remove this altogether if it's not needed

File: debian/jellyfin-ffmpeg.postinst [hunk 2]
```diff
@@ -3,3 +3,9 @@
 # Symlink our old locations to the new locations
 # to aid in transitioning around the 10.3.0 release
 ln -s /usr/lib/jellyfin-ffmpeg /usr/share/jellyfin-ffmpeg
+
+# Symlink AMD VAAPI drivers to the dri folder
+if [ $(uname -m) = "x86_64" ]; then
+	ln -s /usr/lib/x86_64-linux-gnu/dri/r600_drv_video.so /usr/lib/jellyfin-ffmpeg/dri
+	ln -s /usr/lib/x86_64-linux-gnu/dri/radeonsi_drv_video.so /usr/lib/jellyfin-ffmpeg/dri
+fi
```
Comment (JustAMan) [inline debian/jellyfin-ffmpeg.postinst]: why is this needed?

File: debian/rules [hunk 3]
```diff
@@ -10,19 +10,20 @@ ifeq ($(VERSION_SUFFIX),testing)
 endif
 PACKAGEVERSION := "$(ORIG_VERSION)-$(VERSION_SUFFIX)"
 
-CONFIG := --toolchain=hardened \
-	--prefix=/usr \
+CONFIG := --prefix=/usr \
```
Comment (JustAMan) [inline debian/rules]: I would rather we not pollute `/usr`, stuff all our built files into `/usr/lib/jellyfin-ffmpeg`

File: debian/rules [hunk 4]
```diff
@@ -40,23 +41,27 @@ CONFIG := --toolchain=hardened \
...
-	--arch=armhf
+	--cross-prefix=/usr/bin/arm-linux-gnueabihf-
 
-CONFIG_ARM64 := --enable-cross-compile \
-	--cross-prefix=/usr/bin/aarch64-linux-gnu- \
-	--arch=arm64
+CONFIG_ARM64 := --arch=arm64 \
+	--enable-omx \
+	--enable-omx-rpi \
+	--enable-cross-compile \
```
Comment (JustAMan) [inline debian/rules]: maybe extract common ARM flags in a separate variable?..

File: debian/rules [hunk 5]
```diff
@@ -40,23 +41,27 @@ CONFIG := --toolchain=hardened \
...
+	--cross-prefix=/usr/bin/aarch64-linux-gnu-
 
 CONFIG_x86 := --arch=amd64 \
-	--enable-nvenc --enable-nvdec --enable-amf
+	--enable-vaapi \
+	--enable-amf \
+	--enable-nvenc \
+	--enable-nvdec \
+	--extra-ldflags=-Wl,-rpath=/usr/lib/jellyfin-ffmpeg/libs
+#	--enable-libmfx
```
Comment (JustAMan) [inline debian/rules]:
```suggestion
#	--enable-libmfx # uncomment for non-free libva
```

File: docker-build.sh [hunk 6]
```diff
@@ -5,6 +5,123 @@
 set -o errexit
 set -o xtrace
 
+ARCHIVE_ADDR=https://archive.ubuntu.com/ubuntu/
+PORTS_ADDR=https://ports.ubuntu.com/
+
+# Prepare HWA headers, libs and drivers for x86_64-linux-gnu
+prepare_hwa_amd64() {
+    # Download and install the nvidia headers from deb-multimedia
```
Comment (JustAMan) [inline docker-build.sh]: this comment is wrong, you're getting the headers from VLC :)

File: docker-build.sh [hunk 7]
```diff
@@ -5,6 +5,123 @@
...
+    make
+    make install
+    popd
+    popd
+
+    # Download and setup AMD AMF headers from AMD official github repo
+    # https://www.ffmpeg.org/general.html#AMD-AMF_002fVCE
+    svn checkout https://github.com/GPUOpen-LibrariesAndSDKs/AMF/trunk/amf/public/include
+    pushd include
+    mkdir -p /usr/include/AMF && mv * /usr/include/AMF
```
Comment (JustAMan) [inline docker-build.sh]:
```suggestion
    mkdir -p /usr/include/AMF
    mv * /usr/include/AMF
```
this reads cleaner IMHO and separate `&&` is not needed because the script has `errexit` set and `mkdir -p` doesn't return non-zero if folder exists

File: docker-build.sh [hunk 8]
```diff
@@ -5,6 +5,123 @@
...
+    popd
+
+    # Download and install libva
+    pushd ${SOURCE_DIR}
+    git clone -b v2.6-branch --depth=1 https://github.com/intel/libva
+    pushd libva
+    sed -i 's|getenv("LIBVA_DRIVERS_PATH")|"/usr/lib/jellyfin-ffmpeg/dri"|g' va/va.c
+    sed -i 's|getenv("LIBVA_DRIVER_NAME")|NULL|g' va/va.c
```
Comment (JustAMan) [inline docker-build.sh]: why do we need these instead of just exporting those variables somewhere?..

File: docker-build.sh [hunk 9]
```diff
@@ -5,6 +5,123 @@
...
+    ./autogen.sh
+    ./configure --prefix=/usr
```
Comment (JustAMan) [inline docker-build.sh]: `prefix` value doesn't look correct, I think it should be `/usr/lib/jellyfin-ffmpeg`

File: docker-build.sh [hunk 10]
```diff
@@ -5,6 +5,123 @@
...
+    echo "intel-drivers/libva-drm.so* usr/lib/jellyfin-ffmpeg/libs" >> ${SOURCE_DIR}/debian/jellyfin-ffmpeg.install
+    popd
+    popd
+
+    # Download and install intel-vaapi-driver
+    pushd ${SOURCE_DIR}
+    git clone -b v2.4-branch --depth=1 https://github.com/intel/intel-vaapi-driver
+    pushd intel-vaapi-driver
+    ./autogen.sh
+    ./configure --prefix=/usr
```
Comment (JustAMan) [inline docker-build.sh]: same about `prefix`, it should point to the correct location from the get-go, not relying on the fact that it works _for now_ when we move files

File: docker-build.sh [hunk 11]
```diff
@@ -5,6 +5,123 @@
...
+    make -j$(nproc) && make install
+    mkdir -p ${SOURCE_DIR}/intel-drivers/dri
+    cp /usr/lib/dri/i965_drv_video.so ${SOURCE_DIR}/intel-drivers/dri
+    echo "intel-drivers/dri/i965*.so usr/lib/jellyfin-ffmpeg/dri" >> ${SOURCE_DIR}/debian/jellyfin-ffmpeg.install
+    #export LIBVA_DRIVER_NAME=i965
+    #export LIBVA_DRIVERS_PATH=/usr/lib/dri
+    #export PKG_CONFIG_PATH=/usr/lib/pkgconfig
```
Comment (JustAMan) [inline docker-build.sh]:
```suggestion
    # uncomment for non-free libva
    #export LIBVA_DRIVER_NAME=i965
    #export LIBVA_DRIVERS_PATH=/usr/lib/dri
    #export PKG_CONFIG_PATH=/usr/lib/pkgconfig
```

File: docker-build.sh [hunk 12]
```diff
@@ -5,6 +5,123 @@
...
+    popd
+    popd
+
+    # Download and install gmmlib
```
Comment (JustAMan) [inline docker-build.sh:58]:
```suggestion
    # uncomment for non-free libva
    # Download and install gmmlib
```

**Review by JustAMan — CHANGES_REQUESTED** [review]
This also seems to lack the same changes to `Dockerfile` which is used to build `jellyfin-ffmpeg` which we ship within our Docker images.

File: docker-build.sh (hunk 9 above)
Comment (nyanmisaka) [reply docker-build.sh]:
```
Depends:
ffmpeg <---- libva, MediaSDK
intel-vaapi-driver <---- libva
media-driver <---- libva, gmmlib
MediaSDK <---- libva, media-driver
```

You got it wrong. This `/usr` is the path when building `ffmpeg` in docker, in order to allow `/usr/lib` and `/usr/include` to be correctly recognized by `intel-vaapi-driver` and `ffmpeg`, so that they can start building. Once the build is complete, only the files specified in `debian/ jellyfin-ffmpeg.install` will be included in the deb package.

These dockers are all disposable, and can be cleaned up every time the build is completed.

Same about `prefix` in other parts.

File: debian/jellyfin-ffmpeg.postinst (hunk 2 above)
Comment (nyanmisaka) [reply debian/jellyfin-ffmpeg.postinst]: When ffmpeg uses vaapi, it will look for an environment variable `LIBVA_DRIVERS_PATH` (usually located in `/usr/lib/dri`). Since the location of the Intel VAAPI driver in the deb package is `jellyfin-ffmpeg/dri`, I modified this part of `va.c` to make `libva` skip the search for the driver that comes with the system and use the driver included in the deb. Therefore, we also need to symlink the AMD VAAPI drivers here.

If these environment variables are still used, once the user destroys them or makes changes during system update installation, `libva` in the deb package will not find a custom driver path.

File: docker-build.sh (hunk 8 above)
Comment (nyanmisaka) [reply docker-build.sh]: System variables are unstable and can be easily modified during system update or software package uninstallation. Will affect the use of the vaapi driver provided by the official repository of Ubuntu/Debian. And it will cause the `libva` included in the deb package to not find the integrated intel driver. My purpose is to make the dynamic link libraries integrated in the `jellyfin-ffmpeg` deb package coexist with those already existing in those systems. After all, some people still use linux desktop for daily use.

**Comment by nyanmisaka** [comment]
> This also seems to lack the same changes to `Dockerfile` which is used to build `jellyfin-ffmpeg` which we ship within our Docker images.

I am not quite sure why we need to maintain another ffmpeg docker image. When building the Jellyfin image, include this deb package in it and you can use it normally. The test I did for this PR was done in docker.

https://github.com/jellyfin/jellyfin/pull/2813

File: docker-build.sh (hunk 6 above)
Comment (nyanmisaka) [reply docker-build.sh]: Oh. I didn't add this part, I just moved it here ;)

File: debian/jellyfin-ffmpeg.postinst (hunk 2 above)
Comment (JustAMan) [reply debian/jellyfin-ffmpeg.postinst]: I suggest then replacing `LIBVA_DRIVERS_PATH` in the compiled `libva` with `/usr/lib/jellyfin-ffmpeg/dri:/usr/lib/x86_64-linux-gnu/dri/:/usr/lib/dri`, this would make this hack not needed

File: docker-build.sh (hunk 6 above)
Comment (JustAMan) [reply docker-build.sh]: I see, but it doesn't mean you cannot fix it :)

File: docker-build.sh (hunk 8 above)
Comment (JustAMan) [reply docker-build.sh]: I wasn't suggesting to export those globally. Also, I think I am now proposing a better solution: https://github.com/jellyfin/jellyfin-ffmpeg/pull/33#discussion_r406145592 (needs to be checked if it works, though).

File: debian/jellyfin-ffmpeg.postinst (hunk 2 above)
Comment (nyanmisaka) [reply debian/jellyfin-ffmpeg.postinst]: Does this mean that libva will look for `.so` in a colon-separated order?

File: docker-build.sh (hunk 9 above)
Comment (JustAMan) [reply docker-build.sh]: No, I got it right. :)
The fact that stuff works _now_ after you move all files built in docker to some other location is a fortunate happenstance (i.e. it's not guaranteed to happen). You need to make additional actions to ensure your binaries could be shuffled like that, because sometimes the prefix is baked inside the library as part of some path, and then everything break.

So what I'm saying is not that _you_ did it completely wrong (it was already a bit wrong before this PR 😄), but rather I would like to see a cleaner approach. That file was based off `debian/control` which installed the results into `/usr`, so for it the prefix was right. But we aren't installing into `/usr`! I'm only saying `prefix` should really match with where we're installing, otherwise some subtle and hard-to-understand bugs may arise.

File: docker-build.sh (hunk 7 above)
Comment (JustAMan) [reply docker-build.sh]: > I am not quite sure why we need to maintain another ffmpeg docker image.

Because legacy reasons. I think we should use single ffmpeg, too.

**Comment by JustAMan** [comment]
> I am not quite sure why we need to maintain another ffmpeg docker image. When building the Jellyfin image, include this deb package in it and you can use it normally.

Because legacy reasons (that docker build was implemented first). I think we should use single ffmpeg, too.
cc @nvllsvm for more opinions

File: debian/jellyfin-ffmpeg.postinst (hunk 2 above)
Comment (JustAMan) [reply debian/jellyfin-ffmpeg.postinst]: No idea :) But that's how typical `FOO_PATH` stuff works on Linux.
Let me quickly read their source.

Comment (JustAMan) [reply debian/jellyfin-ffmpeg.postinst]: Yes, it works just as I expected, see here and below: https://github.com/intel/libva/blob/master/va/va.c#L442-L458

File: Dockerfile.in [hunk 13]
```diff
@@ -3,15 +3,23 @@ FROM DISTRO
...
+ENV DEB_BUILD_OPTIONS=noddebs
+ENV DEBIAN_FRONTEND=noninteractive
 ENV ARCH=BUILD_ARCHITECTURE
 ENV GCC_VER=GCC_RELEASE_VERSION
 ENV SOURCE_DIR=/ffmpeg
 ENV ARTIFACT_DIR=/dist
-ENV DEB_BUILD_OPTIONS=noddebs
+ENV TARGET_DIR=/usr/lib/jellyfin-ffmpeg
+ENV PKG_CONFIG_PATH=${TARGET_DIR}/lib/pkgconfig:$PKG_CONFIG_PATH
+ENV LD_LIBRARY_PATH=${TARGET_DIR}/lib:${TARGET_DIR}/lib/mfx:${TARGET_DIR}/lib/xorg:${LD_LIBRARY_PATH}
```
Comment (JustAMan) [inline Dockerfile.in:14]: why is this needed?

File: debian/changelog [hunk 14]
```diff
@@ -1,3 +1,9 @@
+jellyfin-ffmpeg (4.2.1-7) unstable; urgency=medium
+
+  * Integrate free libva and intel-vaapi-driver with MIT license in the deb package.
```
Comment (JustAMan) [inline debian/changelog]:
```suggestion
  * Integrate free libva and intel-vaapi-driver with MIT license in the deb package to enable full features of VAAPI transcoding without proprietary drivers (see https://github.com/jellyfin/jellyfin-ffmpeg/pull/33).
```
or something similar to explain not only _what_ you're doing but also _why_

File: debian/rules [hunk 15]
```diff
@@ -40,23 +39,25 @@ CONFIG := --toolchain=hardened \
...
 	--enable-omx-rpi \
-	--enable-version3 \
-	--enable-vaapi \
-	--enable-vdpau \
 
-CONFIG_ARM := --enable-mmal \
-	--enable-cross-compile \
+CONFIG_ARM := --arch=armhf \
+	--enable-mmal \
 	--cross-prefix=/usr/bin/arm-linux-gnueabihf- \
```
Comment (JustAMan) [inline debian/rules]:
```suggestion
CONFIG_ARM := ${CONFIG_ARM_COMMON} \
	--arch=armhf \
	--enable-mmal \
	--cross-prefix=/usr/bin/arm-linux-gnueabihf-
```
so the code further down doesn't care about ARM having some common

File: debian/rules [hunk 16]
```diff
@@ -40,23 +39,25 @@ CONFIG := --toolchain=hardened \
...
-	--arch=armhf
 
-CONFIG_ARM64 := --enable-cross-compile \
+CONFIG_ARM64 := --arch=arm64 \
 	--cross-prefix=/usr/bin/aarch64-linux-gnu- \
```
Comment (JustAMan) [inline debian/rules]:
```suggestion
CONFIG_ARM64 := ${CONFIG_ARM_COMMON} \
	--arch=arm64 \
	--cross-prefix=/usr/bin/aarch64-linux-gnu-
```

File: debian/rules [hunk 17]
```diff
@@ -66,10 +67,12 @@ ifeq ($(BUILD_ARCH),x86_64-linux-gnu)
 endif
 ifeq ($(BUILD_ARCH),arm-linux-gnueabihf)
 	# Cross-building ARM on AMD64
+	CONFIG += $(CONFIG_ARM_COMMON)
```
Comment (JustAMan) [inline debian/rules]: If you take my suggestions above this one won't be needed

**Review by JustAMan — APPROVED** [review]

File: Dockerfile.in (hunk 13 above)
Comment (nyanmisaka) [reply Dockerfile.in:14]: Since the installation location of the libva/qsv header and pkgconfig is modified, we need to manually specify their paths so that they can be found when configure ffmpeg.

**Review by joshuaboniface — APPROVED** [review]

**Comment by JustAMan** [comment]
I think it makes sense to make a follow-up PR to update our building of Docker ffmpeg image - we should add our repo and install `jellyfin-ffmpeg` there instead of duplicating the efforts.

**Comment by Legogris** [comment]
I think part of this should be reverted - VDPAU is required to get VAAPI hardware transcoding working with `amdgpu`.

**Comment by nyanmisaka** [comment]
> I think part of this should be reverted - VDPAU is required to get VAAPI hardware transcoding working with `amdgpu`.

I do not think so. VDPAU is NVIDIA's open source driver video acceleration protocol. AMD GFX can access VAAPI through `radeonsi_drv_video.so`. The two are independent of each other. Because of the rise of NVENC/NVDEC, it is not difficult to see that NVIDIA has almost given up VDPAU. I have tested the availability of VAAPI on Vega 64 and RX 580. One special case is that Mesa 20.1 is required to properly transcode HEVC video.

**Comment by Legogris** [comment]
@nyanmisaka Thanks for following up. I'm raising it because I'm struggling with getting hardware acceleration working on 3400G (Vega 11). 

Jellyfin running in Docker.
Host: Debian buster10, amdgpu kernel drivers from backports
Guest: Base image `linuxserver/jellyfin` (e.g. ubuntu bionic), userspace installed/upgraded from  from `ppa:oibaf/graphics-drivers` (which implies mesa 20.2)

Just in case you're open to troubleshooting and think it's relevant,  
https://github.com/linuxserver/docker-jellyfin/issues/38

**Comment by kushagharahi** [comment]
VDPAU should stay so Jellyfin can run on legacy hardware/nvidia drivers. Just because nvidia does not give it's legacy api much attention doesn't mean this project should quit using it. 

Because VDPAU was removed, I can not get HW acceleration on my Geforce 8200M G.

**Comment by nyanmisaka** [comment]
> VDPAU should stay so Jellyfin can run on legacy hardware/nvidia drivers. Just because nvidia does not give it's legacy api much attention doesn't mean this project should quit using it.
> 
> Because VDPAU was removed, I can not get HW acceleration on my Geforce 8200M G.

I understand this need, will restore VDPAU in the next version but I have no access to such old hardware to ensure it works.

**Comment by nyanmisaka** [comment]
> @nyanmisaka Thanks for following up. I'm raising it because I'm struggling with getting hardware acceleration working on 3400G (Vega 11).
> 
> Jellyfin running in Docker.
> Host: Debian buster10, amdgpu kernel drivers from backports
> Guest: Base image `linuxserver/jellyfin` (e.g. ubuntu bionic), userspace installed/upgraded from from `ppa:oibaf/graphics-drivers` (which implies mesa 20.2)
> 
> Just in case you're open to troubleshooting and think it's relevant,
> [linuxserver/docker-jellyfin#38](https://github.com/linuxserver/docker-jellyfin/issues/38)

Sorry for the late reply. You need to install Mesa 20.1 or higher in the docker that contains Jellyfin so as to let libva detect the new driver.

I'm considering integrating `radeonsi_drv_video.so` into `jellyfin-ffmpeg`, but I want some way to avoid building the whole Mesa.
